// @apibrew/client/nano/model/flow.js
export var FlowEntityInfo = {
    namespace: "nano",
    resource: "Flow",
    restPath: "nano-flow",
};
export var Action;
(function (Action) {
    Action["CREATE"] = "CREATE";
    Action["GET"] = "GET";
    Action["LOAD"] = "LOAD";
    Action["UPDATE"] = "UPDATE";
    Action["DELETE"] = "DELETE";
    Action["LIST"] = "LIST";
})(Action || (Action = {}));
export var Order;
(function (Order) {
    Order["BEFORE"] = "BEFORE";
    Order["AFTER"] = "AFTER";
})(Order || (Order = {}));
export var Kind;
(function (Kind) {
    Kind["ACTION"] = "ACTION";
    Kind["EVENT"] = "EVENT";
    Kind["API_CREATE"] = "API_CREATE";
    Kind["API_GET"] = "API_GET";
    Kind["API_LOAD"] = "API_LOAD";
    Kind["API_UPDATE"] = "API_UPDATE";
    Kind["API_DELETE"] = "API_DELETE";
    Kind["API_LIST"] = "API_LIST";
    Kind["CODE"] = "CODE";
    Kind["FAIL"] = "FAIL";
    Kind["ASSIGN"] = "ASSIGN";
    Kind["CONDITION"] = "CONDITION";
    Kind["HTTP_CALL"] = "HTTP_CALL";
    Kind["FUNCTION_CALL"] = "FUNCTION_CALL";
    Kind["TEMPLATE_CALL"] = "TEMPLATE_CALL";
    Kind["END"] = "END";
})(Kind || (Kind = {}));
export var FlowResource = {
    "auditData": {
        "createdBy": "admin",
        "createdOn": "2024-04-17T10:31:07Z"
    },
    "name": "Flow",
    "namespace": {
        "name": "nano"
    },
    "properties": {
        "id": {
            "type": "UUID",
            "primary": true,
            "required": true,
            "immutable": true,
            "exampleValue": "a39621a4-6d48-11ee-b962-0242ac120002",
            "description": "The unique identifier of the resource. It is randomly generated and immutable.",
            "annotations": {
                "SpecialProperty": "true"
            }
        },
        "name": {
            "type": "STRING",
            "required": true,
            "unique": true,
            "length": 255,
            "title": "Name",
            "description": "Name"
        },
        "statements": {
            "type": "LIST",
            "required": true,
            "item": {
                "type": "STRUCT",
                "typeRef": "Statement"
            },
            "defaultValue": null
        },
        "trigger": {
            "type": "STRING",
            "required": true,
            "length": 255,
            "title": "Trigger",
            "description": "Trigger"
        },
        "version": {
            "type": "INT32",
            "required": true,
            "defaultValue": 1,
            "exampleValue": 1,
            "title": "Version",
            "description": "The version of the resource/record. It is incremented on every update.",
            "annotations": {
                "AllowEmptyPrimitive": "true",
                "SpecialProperty": "true"
            }
        }
    },
    "types": [
        {
            "name": "EventParams",
            "title": "",
            "description": "",
            "properties": {
                "action": {
                    "type": "ENUM",
                    "enumValues": [
                        "CREATE",
                        "GET",
                        "LOAD",
                        "UPDATE",
                        "DELETE",
                        "LIST"
                    ]
                },
                "annotations": {
                    "type": "MAP",
                    "item": {
                        "type": "STRING"
                    }
                },
                "finalizes": {
                    "type": "BOOL"
                },
                "order": {
                    "type": "ENUM",
                    "enumValues": [
                        "BEFORE",
                        "AFTER"
                    ]
                },
                "responds": {
                    "type": "BOOL"
                },
                "sync": {
                    "type": "BOOL"
                },
                "type": {
                    "type": "STRING"
                }
            }
        },
        {
            "name": "ActionParams",
            "title": "",
            "description": "",
            "properties": {
                "type": {
                    "type": "STRING"
                }
            }
        },
        {
            "name": "ApiSaveParams",
            "title": "",
            "description": "",
            "properties": {
                "payload": {
                    "type": "OBJECT"
                },
                "type": {
                    "type": "STRING"
                }
            }
        },
        {
            "name": "ApiLoadParams",
            "title": "",
            "description": "",
            "properties": {
                "match": {
                    "type": "OBJECT"
                },
                "params": {
                    "type": "OBJECT"
                },
                "type": {
                    "type": "STRING"
                }
            }
        },
        {
            "name": "AssignParams",
            "title": "",
            "description": "",
            "properties": {
                "expression": {
                    "type": "STRING"
                },
                "left": {
                    "type": "STRING"
                }
            }
        },
        {
            "name": "FunctionCallParams",
            "title": "",
            "description": "",
            "properties": {
                "name": {
                    "type": "STRING"
                },
                "params": {
                    "type": "OBJECT"
                }
            }
        },
        {
            "name": "CodeParams",
            "title": "",
            "description": "",
            "properties": {
                "content": {
                    "type": "STRING"
                }
            }
        },
        {
            "name": "ConditionParams",
            "title": "",
            "description": "",
            "properties": {
                "condition": {
                    "type": "STRING"
                },
                "fail": {
                    "type": "LIST",
                    "required": true,
                    "item": {
                        "type": "STRUCT",
                        "typeRef": "Statement"
                    },
                    "defaultValue": null
                },
                "pass": {
                    "type": "LIST",
                    "required": true,
                    "item": {
                        "type": "STRUCT",
                        "typeRef": "Statement"
                    },
                    "defaultValue": null
                }
            }
        },
        {
            "name": "FailParams",
            "title": "",
            "description": "",
            "properties": {
                "message": {
                    "type": "STRING"
                }
            }
        },
        {
            "name": "Statement",
            "title": "",
            "description": "",
            "properties": {
                "checkResult": {
                    "type": "BOOL"
                },
                "kind": {
                    "type": "ENUM",
                    "enumValues": [
                        "ACTION",
                        "EVENT",
                        "API_CREATE",
                        "API_GET",
                        "API_LOAD",
                        "API_UPDATE",
                        "API_DELETE",
                        "API_LIST",
                        "CODE",
                        "FAIL",
                        "ASSIGN",
                        "CONDITION",
                        "HTTP_CALL",
                        "FUNCTION_CALL",
                        "TEMPLATE_CALL",
                        "END"
                    ]
                },
                "params": {
                    "type": "OBJECT"
                },
                "variable": {
                    "type": "STRING"
                }
            }
        }
    ]
};
