// @apibrew/client/extension-info.js
export function toExtension(extensionInfo) {
    return {
        name: extensionInfo.namespaces + "/" + extensionInfo.resources + "/" + extensionInfo.action + "/" + extensionInfo.order + "/" + extensionInfo.sync + "/" + extensionInfo.responds + "/" + extensionInfo.finalizer,
        finalizes: Boolean(extensionInfo.finalizer),
        sync: Boolean(extensionInfo.sync),
        responds: Boolean(extensionInfo.responds),
        order: extensionInfo.order,
        selector: prepareEventSelector(extensionInfo),
    };
}
export function prepareEventSelector(extensionInfo) {
    return {
        namespaces: extensionInfo.namespaces,
        resources: extensionInfo.resources,
        actions: [extensionInfo.action],
    };
}
export function extensionInfoToString(extensionInfo) {
    if (!extensionInfo) {
        return '';
    }
    return extensionInfo.namespaces + "/" + extensionInfo.resources + "/" + extensionInfo.action + "/" + extensionInfo.order + "/" + extensionInfo.sync + "/" + extensionInfo.responds + "/" + extensionInfo.finalizer;
}
