// @apibrew/client/util/base-annotations.js
// Constants
var _a;
import { Type } from "../model/resource";
export var Enabled = "true";
// Resource flags
export var KeepHistory = "KeepHistory";
export var HistoryResource = "HistoryResource";
export var NormalizedResource = "NormalizedResource";
export var AutoCreated = "AutoCreated";
export var EnableAudit = "EnableAudit";
export var DisableVersion = "DisableVersion";
export var DisableBackup = "DisableBackup";
// Property params
export var SourceDef = "SourceDef";
export var SourceIdentity = "SourceIdentity";
export var SourceMatchKey = "SourceMatchKey";
export var Identity = "Identity";
export var SpecialProperty = "SpecialProperty";
// HCL
export var IsHclLabel = "IsHclLabel";
export var HclBlock = "HclBlock";
// Request annotations
export var IgnoreIfExists = "IgnoreIfExists";
export var CheckVersion = "CheckVersion";
// Security
export var AllowPublicAccess = "AllowPublicAccess";
export var AllowPublicReadAccess = "AllowPublicGetAccess";
export var AllowPublicCreateAccess = "AllowPublicCreateAccess";
export var AllowPublicUpdateAccess = "AllowPublicUpdateAccess";
export var AllowPublicDeleteAccess = "AllowPublicDeleteAccess";
// SQL
export var SQLType = "SQLType";
export var SQLUseTextType = "SQLUseTextType";
export var UseJoinTable = "UseJoinTable";
// Code generator
export var TypeName = "TypeName";
export var SelfContainedProperty = "SelfContainedProperty";
export var AllowEmptyPrimitive = "AllowEmptyPrimitive";
export var CommonType = "CommonType";
// REST API
export var RestApiDisabled = "RestApiDisabled";
// OpenAPI
export var OpenApiGroup = "OpenApiGroup";
export var OpenApiHide = "OpenApiHide";
export var OpenApiRestPath = "OpenApiRestPath";
// Service
export var ServiceKey = "ServiceKey";
// ExtensionId
export var ExtensionId = "ExtensionId";
// Bypass extensions
export var BypassExtensions = "BypassExtensions";
export var StudioSeparatePages = "StudioSeparatePages";
// Client allowed annotations
export var ClientAllowedAnnotations = (_a = {},
    _a[BypassExtensions] = true,
    _a);
var annotations = [
    { name: Enabled, type: Type.BOOL },
    { name: KeepHistory, type: Type.BOOL },
    { name: HistoryResource, type: Type.BOOL },
    { name: NormalizedResource, type: Type.BOOL },
    { name: AutoCreated, type: Type.BOOL },
    { name: EnableAudit, type: Type.BOOL },
    { name: DisableVersion, type: Type.BOOL },
    { name: DisableBackup, type: Type.BOOL },
    { name: SourceDef, type: Type.BOOL },
    { name: SourceIdentity, type: Type.BOOL },
    { name: SourceMatchKey, type: Type.BOOL },
    { name: Identity, type: Type.BOOL },
    { name: SpecialProperty, type: Type.BOOL },
    { name: IsHclLabel, type: Type.BOOL },
    { name: HclBlock, type: Type.BOOL },
    { name: IgnoreIfExists, type: Type.BOOL },
    { name: CheckVersion, type: Type.BOOL },
    { name: AllowPublicAccess, type: Type.BOOL },
    { name: AllowPublicReadAccess, type: Type.BOOL },
    { name: AllowPublicCreateAccess, type: Type.BOOL },
    { name: AllowPublicUpdateAccess, type: Type.BOOL },
    { name: AllowPublicDeleteAccess, type: Type.BOOL },
    { name: SQLType, type: Type.BOOL },
    { name: SQLUseTextType, type: Type.BOOL },
    { name: UseJoinTable, type: Type.BOOL },
    { name: TypeName, type: Type.BOOL },
    { name: SelfContainedProperty, type: Type.BOOL },
    { name: AllowEmptyPrimitive, type: Type.BOOL },
    { name: CommonType, type: Type.BOOL },
    { name: RestApiDisabled, type: Type.BOOL },
    { name: OpenApiGroup, type: Type.BOOL },
    { name: OpenApiHide, type: Type.BOOL },
    { name: OpenApiRestPath, type: Type.BOOL },
    { name: ServiceKey, type: Type.BOOL },
    { name: ExtensionId, type: Type.BOOL },
    { name: BypassExtensions, type: Type.BOOL },
    { name: StudioSeparatePages, type: Type.BOOL }
];
